vm/rust/src/lib.rs

```diff
@@ -98,6 +109,106 @@ pub extern "C" fn cairoVMCall(
+    txns_json: *const c_char,
+    classes_json: *const c_char,
+    reader_handle: usize,
+    block_number: c_ulonglong,
+    block_timestamp: c_ulonglong,
+    chain_id: *const c_char,
+    sequencer_address: *const c_uchar,
+) {
+    let reader = JunoStateReader::new(reader_handle);
+    let chain_id_str = unsafe { CStr::from_ptr(chain_id) }.to_str().unwrap();
+    let txn_json_str = unsafe { CStr::from_ptr(txns_json) }.to_str().unwrap();
+    let sn_api_txns: Result<Vec<StarknetApiTransaction>, serde_json::Error> =
+        serde_json::from_str(txn_json_str);
+    if sn_api_txns.is_err() {
+        report_error(reader_handle, sn_api_txns.unwrap_err().to_string().as_str());
+        return;
+    }
+
+    let mut classes: Result<Vec<Box<serde_json::value::RawValue>>, serde_json::Error> = Ok(vec![]);
+    if !classes_json.is_null() {
+        let classes_json_str = unsafe { CStr::from_ptr(classes_json) }.to_str().unwrap();
+        classes = serde_json::from_str(classes_json_str);
+    }
+    if classes.is_err() {
+        report_error(reader_handle, classes.unwrap_err().to_string().as_str());
+        return;
+    }
+
+    let sn_api_txns = sn_api_txns.unwrap();
+    let mut classes = classes.unwrap();
+
+    let sequencer_address_felt = ptr_to_felt(sequencer_address);
+    let block_context: BlockContext = build_block_context(
+        chain_id_str,
+        block_number,
+        block_timestamp,
+        sequencer_address_felt,
+    );
+    let mut state = CachedState::new(reader);
+
+    for sn_api_txn in sn_api_txns {
+        let contract_class = match sn_api_txn.clone() {
+            StarknetApiTransaction::Declare(declare) => {
+                if classes.len() == 0 {
+                    report_error(reader_handle, "missing declared class".to_string().as_str());
+                    return;
+                }
+                let class_json_str = classes.remove(0);
+
+                let mut maybe_cc = contract_class_from_json_str(class_json_str.get());
+                if declare.version() == TransactionVersion(2u32.into()) && maybe_cc.is_err() {
+                    // class json could be sierra
+                    maybe_cc = contract_class_from_sierra_json(class_json_str.get())
+                };
+
+                if maybe_cc.is_err() {
+                    report_error(reader_handle, maybe_cc.unwrap_err().to_string().as_str());
+                    return;
+                }
+                Some(maybe_cc.unwrap())
+            }
+            _ => None,
+        };
+
+        let txn = Transaction::from_api(sn_api_txn.clone(), contract_class, None);
+        if txn.is_err() {
+            report_error(reader_handle, txn.unwrap_err().to_string().as_str());
+            return;
+        }
+
+        let res = match txn.unwrap() {
+            Transaction::AccountTransaction(t) => t.execute(&mut state, &block_context),
+            Transaction::L1HandlerTransaction(t) => t.execute(&mut state, &block_context),
+        };
```

Comment:
Why are we matching for `L1HandlerTransaction`?

Reply:
`cairoVMExecute` is a general purpose txn execution entrypoint. so, even tho we dont need it for estimateFee, it supports L1 handlers as well.